### ISSUE TITLE
Colon character in file/directory on NFS share causing exception

### ISSUE THREAD

**Issue by Petermarcu** [dotnet/corefx#26359]
@PFAR commented on [Mon Jan 15 2018](https://github.com/dotnet/core/issues/1220)

# Colon character in file/directory on NFS share causing exception

Some characters are illegal on Windows file system but are perfectly fine on third party file systems. Windows Explorer and CMD prompt have no problems with directory called "Foo:bar", but .NET Core (and .NET 4) generates System.NotSupportedException: The given path's format is not supported.

# General

Create a share to a Linux machine or NAS device (tested with Synology NAS and NFS, on Windows Server install Client for NFS)
Map that share to x:\
In the share create a folder Test. Inside the folder Test create a folder 'Foo:bar', note that this is easiest done natively on the device, either web interface, putty or FTP.

Create a .NET Core console app. 

var info = new DirectoryInfo("x:\test")
var dirs = info.GetDirectories();
//dirs contains now one entry of "Foo:bar"
var name = dirs[0].Name // <== already wrong, name is reported as "bar"
dirs = dirs[0].GetDirectories(); // <== NotSupportedException

In code there is likely a check whether the colon is used anywhere else besides for drive letter and generates an exception. Better approach would be that creating new folders and files keeps this check, but for existing entries it should just work just like in File Explorer and CMD prompt, where you can browse inside the directory without any issues.

---

@PFAR commented on [Mon Jan 15 2018](https://github.com/dotnet/core/issues/1220#issuecomment-357677624)

Just to let you know, when using SMB shares the Synology exposes such folder names as something with an entirely different name (like "FP8AMI~K"), so this can only be tested properly on a Windows Server with Client for NFS

**Comment by wfurt** [comment]
Are you trying to create fork? https://en.wikipedia.org/wiki/Fork_(file_system)
Colon is special character on ntfs @PFAR

**Comment by PFAR** [comment]
@wfurt no, in this project we're integrating a Synology NAS/SAN product which exposes files natively through FTP(s)/SFTP. A large portion of our customers are Mac users and (often) unaware about naming restrictions on Windows filesystem.
We are building a product which incorporates a web view on the files and provide upload/download/rename/delete/search functionality by using Client for NFS to directly access the file system (which uses ANSI/1252, but luckily converts lossless to UFT8 and back, and that's an entirely different story ;-)).

It's when users use the Colon character in files or folders where things start to go wrong. From an OS standpoint though the additional restriction on the Colon on existing file system entries doesn't make any sense and makes it impossible to browse inside such folders using the System.IO namespace.

In fact, within a CMD prompt I can create a folder Foo:bar without any issues or browse into it.
![foobar](https://user-images.githubusercontent.com/9445543/35030929-1d186038-fb61-11e7-9365-d9d2e4776073.png)
 however in File Explorer I can only browse, not create
![foobar2](https://user-images.githubusercontent.com/9445543/35030952-2eaa13be-fb61-11e7-89f3-503618782055.png)

**Comment by wfurt** [comment]
It is certainly possible there is some (incorrect) implicit assumption. @JeremyKuhne may be right person to comment on.

**Comment by JeremyKuhne** [comment]
Thanks for the details, I'll try and look at this as soon as I can. I'm not keen on the colon block we have in place and I'm inclined to remove it for 2.1.

Note that you should be able to work around all of this by using device paths, e.g. prefix your path with `\\?\`. That should work with 4.6.2 and higher and Core 1.1/2.0.

**Comment by danmosemsft** [comment]
@maryamariyan this is removing a check and then updating tests.. @JeremyKuhne can give details

**Comment by JeremyKuhne** [comment]
This block needs removed:

https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/IO/Path.Windows.cs#L55-L71

Tests are in System.Runtime.Extensions tests. @Anipik knows how to make these changes if you haven't done it yet.

**Comment by JeremyKuhne** [comment]
Also remove the wildcard check below it: https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/IO/Path.Windows.cs#L80-L85

**Comment by JeremyKuhne** [comment]
And make the check for invalid characters on https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/IO/Path.Windows.cs#L78 always false.
